src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -53,29 +53,75 @@ void ChunkPruningRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) co
     _apply_to_inputs(node);
     return;
   }
+
   const auto stored_table = std::static_pointer_cast<StoredTableNode>(current_node);
   DebugAssert(stored_table->input_count() == 0, "Stored table nodes should not have inputs.");
 
-  /**
-   * A chain of predicates followed by a stored table node was found.
-   */
-  auto table = Hyrise::get().storage_manager.get_table(stored_table->table_name);
+  if (predicates_for_table_nodes.contains(stored_table)) {
+    predicates_for_table_nodes[stored_table].push_back(predicate_chain);
+  } else {
+    predicates_for_table_nodes[stored_table] = {predicate_chain};
+  }
+}
 
-  std::set<ChunkID> pruned_chunk_ids;
-  for (auto& predicate : predicate_nodes) {
-    auto new_exclusions = _compute_exclude_list(*table, *predicate->predicate(), stored_table);
-    pruned_chunk_ids.insert(new_exclusions.begin(), new_exclusions.end());
+void ChunkPruningRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  // we want to follow chains of predicates across joins
+  if (node->type != LQPNodeType::Predicate && node->type != LQPNodeType::Join) {
+    _apply_to_inputs(node);
+    return;
   }
 
-  // wanted side effect of using sets: pruned_chunk_ids vector is sorted
-  const auto& already_pruned_chunk_ids = stored_table->pruned_chunk_ids();
-  if (!already_pruned_chunk_ids.empty()) {
-    std::vector<ChunkID> intersection;
-    std::set_intersection(already_pruned_chunk_ids.begin(), already_pruned_chunk_ids.end(), pruned_chunk_ids.begin(),
-                          pruned_chunk_ids.end(), std::back_inserter(intersection));
-    stored_table->set_pruned_chunk_ids(intersection);
-  } else {
-    stored_table->set_pruned_chunk_ids(std::vector<ChunkID>(pruned_chunk_ids.begin(), pruned_chunk_ids.end()));
+  // Gather PredicateNodes on top of a StoredTableNode. Ignore non-filtering, ValidateNodes and Joins.
+  // Starting from the current node, drilling down might result in multple cains of predicates ending
+  // at different or the same StoredTableNodes.
+  // predicates_for_table_nodes keeps track of the predicate chains for each stored table
+  std::unordered_map<std::shared_ptr<StoredTableNode>, std::vector<PredicateChain>> predicates_for_table_nodes{};
```

Comment:
Why do we need to hold multiple predicate chains? Is there a semantic difference between a single chain and a set of subchains?

Reply:
did some refactoring but the PredicateChain stayed. The motivation here is that a table can have multiple outputs and thus have multiple chains of predicates. for each chain, a seperate exclude list is calculated and finally the intersection of all exclude lists is pruned.